Added an encode routine for a CanId

ad3154 opened the pull request:
* Added a way to encode an extended CAN ID from its components. 
* Added an alias "BROADCAST" address which is the same as "GLOBAL". Internally at Raven we call it global but more broadly (and in AgIsoStack++) it seems to be called "broadcast".

ad3154 commented on src/driver/can_id.rs:
```diff
@@ -71,6 +71,34 @@ impl CanId {
+        source_address: Address,
+        destination_address: Address,
+        priority: Priority,
+    ) -> Result<CanId, &'static str> {
```
Is this normal, or should I return `Err<CanID>(0)` or something else entirely? In AgIsoStack++ we log this error because it's usually someone doesn't know what they are doing

Notgnoshi replied:
It's most normal to use an error enum, but this fine too.

I really like embedding the information that caused the fault in the error, or at least the log message, so I'd maybe do something like this:
```rust
#[derive(Debug, Clone)]
pub struct EncodingError{p: Priority, pgn: Pgn, sa: Address, da: Address}

pub fn try_encode(pgn: Pgn, sa: Address, da: Address, p: Priority) -> Result<CanId, EncodingError>;
pub unsafe fn encode_unchecked(...) -> CanId;
```
And then the user can do

```rust
match try_encode(...) {
    Ok(id) => { ... },
    Err(e) => log::error!("Failed to encode CAN ID: {e:?}"),
}
```

In this particular case though, I'm not sure that it's worth it? The way it's implemented right now,

```rust
let id = encode(...).unwrap();
```
will give the right error message in the console, but it won't include the particular address/pgn combo that caused the fault.

ad3154 replied:
I have changed it to match your suggestion because I liked the idea of being able to print better log info, though this comes at the expense of a temporary warning due to not reading the values out of the enum. Since we decided in #15 that warnings were preferable to forcing the pipeline to pass, I think it should still be OK to merge it in.

Notgnoshi commented on src/driver/can_id.rs:
```diff
@@ -71,6 +79,71 @@ impl CanId {
     }
 
+    /// Encodes a new extended ID using the discrete parts of an identifier
+    pub fn try_encode(
+        parameter_group_number: Pgn,
+        source_address: Address,
+        destination_address: Address,
+        priority: Priority,
+    ) -> Result<CanId, EncodingError> {
```
nitpick suggestion:

```rust
pub fn try_encode(
  parameter_group_number: Pgn,
  source_address: Address,
  destination_address: Address,
  priority: Priority,
) -> Result<CanId, EncodingError> {
  if destination_address != Address::GLOBAL && pgn.is_broadcast() {
    return Err(EncodingError { p: priority, pgn: parameter_group_number, sa: source_address, da: destination_address });
  }
  Ok(unsafe { encode_unchecked(parameter_group_number, source_address, destination_address, priority) })
}
```

Notgnoshi reviewed (approved):
offtopic: I like to use https://conventionalcomments.org/ for comments on PRs, because I'm _terrible_ at conveying tone through comments.

When I leave a `nitpick:` comment, I intend to convey that it's not something major, and am comfortable merging the PR without addressing the comment.

Notgnoshi commented on src/driver/can_id.rs:
```diff
@@ -49,6 +49,14 @@ pub enum Type {
 
+#[derive(Debug, Clone)]
+pub struct EncodingError {
+    p: Priority,
+    pgn: Pgn,
+    sa: Address,
+    da: Address,
```
suggestion: To fix the pipeline, these fields should become `pub`.

nitpick: I think these variables should be spelled out. Apologies, when I made the suggestion comment earlier, I was trying to save on typing :cry:

ad3154 commented:
> When I leave a `nitpick:` comment, I intend to convey that it's not something major, and am comfortable merging the PR without addressing the comment.

Yeah, I know 😁 but these are super minor things in this case, I might as well make it as nice as possible now since they're quick, and it helps me learn rust-ways-of-doing-things (which as you all can probably see is something I still need to work on).
